In [1]:
from sklearn import datasets, linear_model
import pandas as pd
import numpy as np

# 기존모델(original.csv, allfeature3.csv, logall3.csv)

In [3]:
df3 = pd.read_csv("original.csv", encoding="euc-kr", index_col="year")

In [4]:
df3

전년도재고량(천톤)  쌀생산량-정부매입량  전년도단경기가격    수확기
year                                         
2005         850     5519676     34931  36303
2006         832     5299798     39843  37501
2007         830     5033547     38858  38222
2008         695     5497138     39510  40591
2009         686     5552768     41842  34562
2010         993     4836019     37427  34081
2011        1509     4774874     32359  40981
2012        1051     4564860     39000  43386
2013         762     4824711     41917  43899
2014         801     4827352     44400  41578
2015         874     4845310     42569  37527
2016        1354     4705895     39892  31905
2017        1747     4465272     35449  38101

In [7]:
# 로그씌우기
lencol = len(df3.columns)
for i in range(lencol):
    l1 = df3.iloc[:, i]
    lenl1 = len(l1)
    l2 = []
    
    for a in l1:
        x = np.log(a)
        l2.append(x)
        
    df3.iloc[:, i] = l2

In [8]:
#모델 돌리기
model1 = linear_model.LinearRegression()
x_vars1 = df3.iloc[:, :-1]
lr = model1.fit(x_vars1, df3.iloc[:, -1])
print(model1.coef_, model1.intercept_)

orir = lr.score(x_vars1, df3.iloc[:, -1])

[-0.49924378 -1.63910527 -0.85873579] 48.32587367015266


In [9]:
#R제곱값
orir = lr.score(x_vars1, df3.iloc[:, -1])
orir

0.43400666468152205

# 변수+1 

In [10]:
df4 = pd.read_csv("allfeature3.csv", encoding="euc-kr", index_col="year")

In [11]:
df4

추계인구(명)  농가인구(명)   추계인구_전국  농가인구_전국  유통비_직접비  유통비_간접비  유통비_이윤  유통비_계  \
year                                                                          
2005  48184561  3433573  48184561  3433573     12.6      6.7     3.9   23.2   
2006  48438292  3304175  48438292  3304175     13.6      5.1     3.6   22.3   
2007  48683638  3274089  48683638  3274089     13.0      5.1     3.1   21.2   
2008  49054708  3186752  49054708  3186752     13.3      4.8     2.8   20.9   
2009  49307835  3117323  49307835  3117323     16.2      1.9     1.5   19.6   
2010  49554112  3062956  49554112  3062956     16.9      2.2     1.6   20.7   
2011  49936638  2962113  49936638  2962113     13.4      4.7     3.2   21.3   
2012  50199853  2911541  50199853  2911541     13.3      5.5     2.0   20.8   
2013  50428893  2847435  50428893  2847435     13.3      5.7     2.5   21.5   
2014  50746659  2751791  50746659  2751791     14.0      5.5     2.5   22.0   
2015  51014947  2569387  51014947  2569387     15.1      6.7     4.0   25.8   
2016  51245707  2496406  51245707  2496406     17.4      7.0     4.3   28.7   
2017  51446201  2422255  51446201  2422255     17.5      7.1     4.4   28.8   

      유통비_출하  유통비_도매  ...      합계면적    합계생산량  1인당 국내총생산(명목 달러표시) (달러)  \
year                  ...                                               
2005    11.4     3.8  ...    979717  4768368                  18636.0   
2006    12.3     4.7  ...    955229  4679991                  20872.7   
2007    11.5     4.7  ...    950250  4407743                  23062.1   
2008    11.6     3.6  ...    935766  4843478                  20420.6   
2009    12.3     4.4  ...    924471  4916080                  18299.5   
2010    13.3     4.5  ...    892074  4295413                  22083.2   
2011    11.5     4.4  ...    853823  4224019                  24083.7   
2012    10.9     3.9  ...    849172  4006183                  24349.9   
2013    11.0     3.8  ...    832625  4230011                  25885.6   
2014    11.9     4.0  ...    815506  4240739                  27804.5   
2015    12.6     5.0  ...    799344  4326915                  27097.1   
2016    14.7     5.4  ...    778734  4196691                  27606.7   
2017    14.8     5.5  ...    754713  3972468                  29743.5   

      1인당 국민총소득(명목 달러표시) (달러)  1인당 가계총처분가능소득(명목 달러표시) (달러)  국내총생산(실질성장률) (%)  \
year                                                                           
2005                  18490.4                      11035.2               3.9   
2006                  20794.7                      12325.2               5.2   
2007                  22992.1                      13236.1               5.5   
2008                  20419.1                      11782.5               2.8   
2009                  18256.2                      10508.9               0.7   
2010                  22105.3                      12234.8               6.5   
2011                  24225.6                      13404.1               3.7   
2012                  24599.9                      13616.9               2.3   
2013                  26070.3                      14643.4               2.9   
2014                  27892.2                      15820.4               3.3   
2015                  27170.9                      15487.4               2.8   
2016                  27681.0                      15514.9               2.9   
2017                  29744.6                      16573.4               3.1   

      쌀생산량-정부매입량  전년도단경기가격    전년도계절진폭    수확기  
year                                          
2005     5519676     34931   8.635204  36303  
2006     5299798     39843  19.800000  37501  
2007     5033547     38858  13.600000  38222  
2008     5497138     39510  13.400000  40591  
2009     5552768     41842  13.100000  34562  
2010     4836019     37427  18.300000  34081  
2011     4774874     32359   4.900000  40981  
2012     4564860     39000   5.200000  43386  
2013     4824711     41917   6.600000  43899  
2014   

In [12]:
#로그처리
lencol = len(df4.columns)
for i in range(lencol):
    l1 = df4.iloc[:, i]
    lenl1 = len(l1)
    l2 = []
    
    for a in l1:
        x = np.log(a)
        l2.append(x)
        
    df4.iloc[:, i] = l2

C:\Anaconda\lib\site-packages\ipykernel_launcher.py:9: RuntimeWarning: divide by zero encountered in log
  if __name__ == '__main__':


In [14]:
df4.iloc[:, 0]

year
2005    17.690549
2006    17.695801
2007    17.700854
2008    17.708447
2009    17.713594
2010    17.718576
2011    17.726266
2012    17.731523
2013    17.736075
2014    17.742356
2015    17.747629
2016    17.752142
2017    17.756047
Name: 추계인구(명), dtype: float64

### logal3.csv는 위에서 log으로 변환한 값들을 넣은 csv파일 + inf값을 0으로 다시 전환한 파일이다. 
- df4.to_csv("logall3.csv", sep = ",", encoding="euc-kr")
- 하지만 0을 log값으로 변환하면 infite로 무한한 값으로 변경되어 에러가 발생될 확률이 높다.
- 이를 위해 inf 값들을 0으로 다시 일일히 처리해야한다.

In [21]:
df5 = pd.read_csv("logall3.csv", encoding="euc-kr", index_col="year")

In [22]:
df5.columns

Index(['추계인구(명)', '농가인구(명)', '추계인구_전국', '농가인구_전국', '유통비_직접비', '유통비_간접비',
       '유통비_이윤', '유통비_계', '유통비_출하', '유통비_도매', '유통비_소매', '자급도(%)',
       '전년도양곡소비량_전가구 (kg)', '전년도소비자물가변동율', '가계지출(소비지출+비소비지출) (천원)', '전년도양곡수요',
       '생산량(톤)', '전년도재고량(천톤)', '최저임금_시간', '최저임금인상율', '직접노동투하량', 'korexpwgt',
       'korimpwgt', '농업소득(천원)', '농가부채(천원)', '논경지면적(ha)', '임차농지비율(%)',
       '임차료율(%)', '농가수', '전년도지급면적', '전년도지급단가', '전년도ha당 단가', '전년도지급총액', '목표가격',
       '농가교역조건지수(A/B)', '농가판매가격지수(A)', '농가구입가격지수(B)', '농업소득 (천원)', '농업소득률 (%)',
       '농업소득의 가계비충족 (%)', '농업순생산 (천원)', '농업의존도 (%)', '부채 (천원)', '자산 (천원)',
       '매입량(톤)', '합계면적', '합계생산량', '1인당 국내총생산(명목 달러표시) (달러)',
       '1인당 국민총소득(명목 달러표시) (달러)', '1인당 가계총처분가능소득(명목 달러표시) (달러)',
       '국내총생산(실질성장률) (%)', '쌀생산량-정부매입량', '전년도단경기가격', '전년도계절진폭', '수확기'],
      dtype='object')

# 회귀분석
- oricols = [ '쌀생산량-정부매입량','전년도단경기가격', '전년도재고량(천톤)']
- 위의 oricols가 정부에서 중요하게 생각하는 변수 값이다.

In [23]:
colnames = ['추계인구(명)', '농가인구(명)', '추계인구_전국', '농가인구_전국', '유통비_직접비', '유통비_간접비',
       '유통비_이윤', '유통비_계', '유통비_출하', '유통비_도매', '유통비_소매', '자급도(%)',
       '전년도양곡소비량_전가구 (kg)', '전년도소비자물가변동율', '가계지출(소비지출+비소비지출) (천원)', '전년도양곡수요',
       '생산량(톤)', '최저임금_시간', '최저임금인상율', '직접노동투하량', 'korexpwgt',
       'korimpwgt', '농업소득(천원)', '농가부채(천원)', '논경지면적(ha)', '임차농지비율(%)',
       '임차료율(%)', '농가수', '전년도지급면적', '전년도지급단가', '전년도ha당 단가', '전년도지급총액', '목표가격',
       '농가교역조건지수(A/B)', '농가판매가격지수(A)', '농가구입가격지수(B)', '농업소득 (천원)', '농업소득률 (%)',
       '농업소득의 가계비충족 (%)', '농업순생산 (천원)', '농업의존도 (%)', '부채 (천원)', '자산 (천원)',
       '매입량(톤)', '합계면적', '합계생산량', '1인당 국내총생산(명목 달러표시) (달러)',
       '1인당 국민총소득(명목 달러표시) (달러)', '1인당 가계총처분가능소득(명목 달러표시) (달러)',
       '국내총생산(실질성장률) (%)','전년도계절진폭']

oricols = [ '쌀생산량-정부매입량','전년도단경기가격', '전년도재고량(천톤)']

#사용할 컬럼값 처리

### '쌀생산량-정부매입량','전년도단경기가격', '전년도재고량(천톤)' 를 기준으로 feature값들을 하나씩 대입해보면서 R값을 산출해본다.

In [24]:
#모델 돌리기
idx = []
r2l = []
r3l =[]
for i in colnames:
    oricols = [ '쌀생산량-정부매입량','전년도단경기가격', '전년도재고량(천톤)']
    oricols.append(i)
    model2 = linear_model.LinearRegression()
    x_vars2 = df5[oricols]
    lr = model2.fit(x_vars2, df5["수확기"])
    r2 = lr.score(x_vars2, df5["수확기"])
    r3 = r2-orir
    if r3 > 0:
        idx.append(i)
        r2l.append(r2)
        r3l.append(r3)
    print(i, r2, r3)

추계인구(명) 0.46395255855995565 0.029945893878433605
농가인구(명) 0.4463887649920084 0.012382100310486366
추계인구_전국 0.46395255855995565 0.029945893878433605
농가인구_전국 0.4463887649920084 0.012382100310486366
유통비_직접비 0.7169288263485387 0.28292216166701667
유통비_간접비 0.5409047067474101 0.10689804206588804
유통비_이윤 0.47248292471200676 0.03847626003048471
유통비_계 0.44305372132709303 0.009047056645570983
유통비_출하 0.7342015589463826 0.30019489426486057
유통비_도매 0.625330189371107 0.1913235246895849
유통비_소매 0.52383896477529 0.08983230009376797
자급도(%) 0.5418795061914703 0.10787284150994825
전년도양곡소비량_전가구 (kg) 0.4418478174502185 0.007841152768696458
전년도소비자물가변동율 0.43714944060792443 0.003142775926402386
가계지출(소비지출+비소비지출) (천원) 0.4829851666146977 0.048978501933175655
전년도양곡수요 0.6379077380804254 0.2039010733989034
생산량(톤) 0.5030002868890879 0.06899362220756589
최저임금_시간 0.4475561500207851 0.013549485339263034
최저임금인상율 0.45419649692139996 0.020189832239877914
직접노동투하량 0.4822463586216569 0.04823969394013483
korexpwgt 0.4654948143702703 

In [25]:
#원래모델-r값 오류땜에 추가
model1 = linear_model.LinearRegression()
x_vars1 = df3.iloc[:, :-1]
lr = model1.fit(x_vars1, df3.iloc[:, -1])
print(model1.coef_, model1.intercept_)

orir = lr.score(x_vars1, df3.iloc[:, -1])
orir

[-0.49924378 -1.63910527 -0.85873579] 48.32587367015266


0.43400666468152205

In [26]:
uplist = pd.DataFrame({"항목": idx, "R2": r2l, "차이": r3l})

In [27]:
#R스퀘어값 변화추이
uplist.sort_values("차이", ascending=False)

항목        R2        차이
8                        유통비_출하  0.734202  0.300195
4                       유통비_직접비  0.716929  0.282922
15                      전년도양곡수요  0.637908  0.203901
9                        유통비_도매  0.625330  0.191324
50                      전년도계절진폭  0.573891  0.139884
34                  농가판매가격지수(A)  0.546375  0.112368
22                     농업소득(천원)  0.545430  0.111424
47      1인당 국민총소득(명목 달러표시) (달러)  0.544370  0.110363
11                       자급도(%)  0.541880  0.107873
5                       유통비_간접비  0.540905  0.106898
48  1인당 가계총처분가능소득(명목 달러표시) (달러)  0.536733  0.102726
46      1인당 국내총생산(명목 달러표시) (달러)  0.530815  0.096808
43                       매입량(톤)  0.528608  0.094602
10                       유통비_소매  0.523839  0.089832
23                     농가부채(천원)  0.505801  0.071794
42                      자산 (천원)  0.504462  0.070455
16                       생산량(톤)  0.503000  0.068994
39                   농업순생산 (천원)  0.500733  0.066726
44                         합계면적  0.484832  0.050826
14        가계지출(소비지출+비소비지출) (천원)  0.482985  0.048979
19                      직접노동투하량  0.482246  0.048240
45                        합계생산량  0.481762  0.047755
49             국내총생산(실질성장률) (%)  0.474261  0.040255
6                        유통비_이윤  0.472483  0.038476
26                      임차료율(%)  0.466033  0.032027
20                    korexpwgt  0.465495  0.031488
29                      전년도지급단가  0.465002  0.030996
30                    전년도ha당 단가  0.464881  0.030874
0                       추계인구(명)  0.463953  0.029946
2                       추계인구_전국  0.463953  0.029946
31                      전년도지급총액  0.463616  0.029609
35                  농가구입가격지수(B)  0.457169  0.023163
18                      최저임금인상율  0.454196  0.020190
36                    농업소득 (천원)  0.452813  0.018806
32                         목표가격  0.450491  0.016485
17                      최저임금_시간  0.447556  0.013549
28                      전년도지급면적  0.446975  0.012969
40                    농업의존도 (%)  0.446745  0.012738
1                       농가인구(명)  0.446389  0.012382
3                       농가인구_전국  0.446389  0.012382
7                         유통비_계  0.443054  0.009047
38              농업소득의 가계비충족 (%)  0.442682  0.008675
12            전년도양곡소비량_전가구 (kg)  0.441848  0.007841
33                농가교역조건지수(A/B)  0.441042  0.007035
24                    논경지면적(ha)  0.439905  0.005899
13                  전년도소비자물가변동율  0.437149  0.003143
21                    korimpwgt  0.436423  0.002416
27                          농가수  0.435637  0.001630
41                      부채 (천원)  0.434781  0.000775
37                    농업소득률 (%)  0.434547  0.000540
25                    임차농지비율(%)  0.434527  0.000520

## 이제 모든 코드를 총합하여 R값 구해보기
- oricols = [ '쌀생산량-정부매입량','전년도단경기가격', '전년도재고량(천톤)', "유통비_출하"]
- 위 값을 기준으로 feature값을 하나씩 대입해보고, 가장 유의미한 R 값들의 목록을 산출해본다.


In [29]:
idx2 = []
r2l2 = []
r3l2 =[]

colnames2 = ['추계인구(명)',
 '농가인구(명)',
 '추계인구_전국',
 '농가인구_전국',
 '자급도(%)',
 '전년도양곡소비량_전가구 (kg)',
 '전년도소비자물가변동율',
 '가계지출(소비지출+비소비지출) (천원)',
 '전년도양곡수요',
 '생산량(톤)',
 '최저임금_시간',
 '최저임금인상율',
 '직접노동투하량',
 'korexpwgt',
 'korimpwgt',
 '농업소득(천원)',
 '농가부채(천원)',
 '논경지면적(ha)',
 '임차농지비율(%)',
 '임차료율(%)',
 '농가수',
 '전년도지급면적',
 '전년도지급단가',
 '전년도ha당 단가',
 '전년도지급총액',
 '목표가격',
 '농가교역조건지수(A/B)',
 '농가판매가격지수(A)',
 '농가구입가격지수(B)',
 '농업소득 (천원)',
 '농업소득률 (%)',
 '농업소득의 가계비충족 (%)',
 '농업순생산 (천원)',
 '농업의존도 (%)',
 '부채 (천원)',
 '자산 (천원)',
 '매입량(톤)',
 '합계면적',
 '합계생산량',
 '1인당 국내총생산(명목 달러표시) (달러)',
 '1인당 국민총소득(명목 달러표시) (달러)',
 '1인당 가계총처분가능소득(명목 달러표시) (달러)',
 '국내총생산(실질성장률) (%)',
 '전년도계절진폭']

for i in colnames2:
    oricols = [ '쌀생산량-정부매입량','전년도단경기가격', '전년도재고량(천톤)', "유통비_출하"]
    oricols.append(i)
    model2 = linear_model.LinearRegression()
    x_vars2 = df5[oricols]
    lr = model2.fit(x_vars2, df5["수확기"])
    r2 = lr.score(x_vars2, df5["수확기"])
    r3 = r2-orir
    if r3 > 0:
        idx2.append(i)
        r2l2.append(r2)
        r3l2.append(r3)
    print(i, r2, r3)
    
uplist2 = pd.DataFrame({"항목": idx2, "R2": r2l2, "차이": r3l2})
uplist2.sort_values("차이", ascending=False)

추계인구(명) 0.7576769680093016 0.3236703033277796
농가인구(명) 0.7676057154016526 0.3335990507201305
추계인구_전국 0.7576769680093016 0.3236703033277796
농가인구_전국 0.7676057154016526 0.3335990507201305
자급도(%) 0.734274443142623 0.300267778461101
전년도양곡소비량_전가구 (kg) 0.7608418019996229 0.3268351373181009
전년도소비자물가변동율 0.7497299873250372 0.3157233226435151
가계지출(소비지출+비소비지출) (천원) 0.8447153036210902 0.4107086389395681
전년도양곡수요 0.7916024380203316 0.35759577333880954
생산량(톤) 0.7344957086046603 0.30048904392313824
최저임금_시간 0.7640293210751601 0.33002265639363804
최저임금인상율 0.743056805240734 0.309050140559212
직접노동투하량 0.7597983939705613 0.3257917292890392
korexpwgt 0.7532345398305591 0.31922787514903705
korimpwgt 0.7353795920421478 0.3013729273606257
농업소득(천원) 0.735906111002322 0.3018994463207999
농가부채(천원) 0.7477439404070902 0.3137372757255682
논경지면적(ha) 0.7536055508543277 0.31959888617280563
임차농지비율(%) 0.7378235724757002 0.30381690779417814
임차료율(%) 0.7348331605548174 0.30082649587329535
농가수 0.7360204654627788 0.3020138007812567


항목        R2        차이
7         가계지출(소비지출+비소비지출) (천원)  0.844715  0.410709
32                   농업순생산 (천원)  0.821691  0.387684
40      1인당 국민총소득(명목 달러표시) (달러)  0.804471  0.370464
39      1인당 국내총생산(명목 달러표시) (달러)  0.802520  0.368514
41  1인당 가계총처분가능소득(명목 달러표시) (달러)  0.796566  0.362559
8                       전년도양곡수요  0.791602  0.357596
29                    농업소득 (천원)  0.790817  0.356810
43                      전년도계절진폭  0.778515  0.344508
37                         합계면적  0.774809  0.340802
35                      자산 (천원)  0.771650  0.337643
1                       농가인구(명)  0.767606  0.333599
3                       농가인구_전국  0.767606  0.333599
25                         목표가격  0.766361  0.332354
31              농업소득의 가계비충족 (%)  0.765477  0.331470
10                      최저임금_시간  0.764029  0.330023
5             전년도양곡소비량_전가구 (kg)  0.760842  0.326835
12                      직접노동투하량  0.759798  0.325792
0                       추계인구(명)  0.757677  0.323670
2                       추계인구_전국  0.757677  0.323670
17                    논경지면적(ha)  0.753606  0.319599
13                    korexpwgt  0.753235  0.319228
6                   전년도소비자물가변동율  0.749730  0.315723
30                    농업소득률 (%)  0.749305  0.315298
16                     농가부채(천원)  0.747744  0.313737
21                      전년도지급면적  0.747325  0.313318
27                  농가판매가격지수(A)  0.745642  0.311635
28                  농가구입가격지수(B)  0.744323  0.310317
11                      최저임금인상율  0.743057  0.309050
33                    농업의존도 (%)  0.741814  0.307808
24                      전년도지급총액  0.740701  0.306695
22                      전년도지급단가  0.740407  0.306400
23                    전년도ha당 단가  0.739783  0.305777
18                    임차농지비율(%)  0.737824  0.303817
26                농가교역조건지수(A/B)  0.736862  0.302855
34                      부채 (천원)  0.736679  0.302672
36                       매입량(톤)  0.736567  0.302561
20                          농가수  0.736020  0.302014
15                     농업소득(천원)  0.735906  0.301899
14                    korimpwgt  0.735380  0.301373
19                      임차료율(%)  0.734833  0.300826
9                        생산량(톤)  0.734496  0.300489
38                        합계생산량  0.734321  0.300314
42             국내총생산(실질성장률) (%)  0.734280  0.300273
4                        자급도(%)  0.734274  0.300268

## 가장 R값이 높은 유의미한 값을 하나씩 넣어서 가장 높은 score 값 산출하기
- feature값들 중 중복되는 데이터로 예상되는 인구, 최저임금 및 수요 등을 빼고 산출하였음

In [97]:
import copy
# 인구가 겹치므로 추계인구(명)을 제외한 나머지 인구 데이터는 삭제
# 또, 최저임금 및 수요를 빼보자.  '최저임금인상율',  '전년도양곡수요', 뺌
colnames2 = ['추계인구(명)',
 '자급도(%)',
 '전년도양곡소비량_전가구 (kg)',
 '전년도소비자물가변동율',
 '가계지출(소비지출+비소비지출) (천원)',
 '생산량(톤)',
 '최저임금_시간',
 '직접노동투하량',
 'korexpwgt',
 'korimpwgt',
 '농업소득(천원)',
 '농가부채(천원)',
 '논경지면적(ha)',
 '임차농지비율(%)',
 '임차료율(%)',
 '농가수',
 '전년도지급면적',
 '전년도지급단가',
 '전년도ha당 단가',
 '전년도지급총액',
 '목표가격',
 '농가교역조건지수(A/B)',
 '농가판매가격지수(A)',
 '농가구입가격지수(B)',
 '농업소득 (천원)',
 '농업소득률 (%)',
 '농업소득의 가계비충족 (%)',
 '농업순생산 (천원)',
 '농업의존도 (%)',
 '부채 (천원)',
 '자산 (천원)',
 '매입량(톤)',
 '합계면적',
 '합계생산량',
 '1인당 국내총생산(명목 달러표시) (달러)',
 '1인당 국민총소득(명목 달러표시) (달러)',
 '1인당 가계총처분가능소득(명목 달러표시) (달러)',
 '국내총생산(실질성장률) (%)',
 '전년도계절진폭']



oricols_origin = [ '쌀생산량-정부매입량','전년도단경기가격', '전년도재고량(천톤)', "유통비_출하"]
oricols = [ '쌀생산량-정부매입량','전년도단경기가격', '전년도재고량(천톤)', "유통비_출하"]
flag = 0
count = 1
while(True) : 
    print(count)
    print("만큼 돌았다.")
    count += 1
    idx2 = []
    r2l2 = []
    r3l2 =[]
    
    for i in colnames2 :
        oricols = copy.deepcopy(oricols_origin)
        oricols.append(i)
        model2 = linear_model.LinearRegression()
        x_vars2 = df5[oricols]
        lr = model2.fit(x_vars2, df5["수확기"])
        r2 = lr.score(x_vars2, df5["수확기"])
        r3 = r2-orir
        if r3 > 0:
            idx2.append(i)
            r2l2.append(r2)
            r3l2.append(r3)
#         print(i, r2, r3)
    
    print("================================================")
    uplist2 = pd.DataFrame({"항목": idx2, "R2": r2l2, "차이": r3l2})
    uplist2.sort_values("차이", ascending=False)
    oricols_origin.append(uplist2['항목'][0])
    colnames2.remove(uplist2['항목'][0])
    if(r2>flag) :
        flag = r2
        print(flag, r2)
    else :
        print(uplist2)
        break

1
만큼 돌았다.
0.7785149502120554 0.7785149502120554
2
만큼 돌았다.
0.9454142104427619 0.9454142104427619
3
만큼 돌았다.
0.9595387319175391 0.9595387319175391
4
만큼 돌았다.
0.973179535387878 0.973179535387878
5
만큼 돌았다.
0.9861297610497455 0.9861297610497455
6
만큼 돌았다.
0.999360335170726 0.999360335170726
7
만큼 돌았다.
0.9995047193340811 0.9995047193340811
8
만큼 돌았다.
1.0 1.0
9
만큼 돌았다.
                             항목   R2        차이
0                     korexpwgt  1.0  0.565993
1                     korimpwgt  1.0  0.565993
2                      농업소득(천원)  1.0  0.565993
3                      농가부채(천원)  1.0  0.565993
4                     논경지면적(ha)  1.0  0.565993
5                     임차농지비율(%)  1.0  0.565993
6                       임차료율(%)  1.0  0.565993
7                           농가수  1.0  0.565993
8                       전년도지급면적  1.0  0.565993
9                       전년도지급단가  1.0  0.565993
10                    전년도ha당 단가  1.0  0.565993
11                      전년도지급총액  1.0  0.565993
12                         목표

### 최종 유의미한 feature 값들은 oricols_origin에 담긴다.
- 마지막에 담긴 'korexpwgt'부터는 1.0으로 고정된 score값을 산출하므로 제외한다.
- 최종적으로 '직접노동투하량'을 포함한 score값은 0.9995 이다.
- 지나치게 높은 score 값들은 과적합을 의심해볼 수 있다.
    - 이를 위해 직접노동투하량 -> 최저임금_시간 -> 생산량(톤) 순으로 feature 값들을 제외한 최종 식을 도출해보는 것도 좋은 방법이다.
- coef는 feature값들의 계수값이고, intercept_ 는 y 절편 값이다.
- 최종적인 식은 아래와 같다.
- *** y = 10.976 * '쌀생산량-정부매입량' + -0.613 * '전년도단경기가격' + ... + 0.318 * '직접노동투하량' - 400.0301 **

-  이 식이 최종적으로 옳은지 검증하기 위해서는 2018년의 값을 집어넣어보고 2018 목표값과 일치하는지 확인해봐야한다. (또는 2017년도 이전)

In [98]:
oricols_origin

['쌀생산량-정부매입량',
 '전년도단경기가격',
 '전년도재고량(천톤)',
 '유통비_출하',
 '추계인구(명)',
 '자급도(%)',
 '전년도양곡소비량_전가구 (kg)',
 '전년도소비자물가변동율',
 '가계지출(소비지출+비소비지출) (천원)',
 '생산량(톤)',
 '최저임금_시간',
 '직접노동투하량',
 'korexpwgt']

In [99]:
# coef는 feature값들의 계수값이고, intercept_ 는 y 절편 값이다.
print(model2.coef_, model2.intercept_)

[ 10.97596495  -0.61303434   0.35312711  -2.20951915  22.06779079
  -0.87833813   3.81275202  -0.05186115   0.43061514 -10.156336
  -0.15429083   0.05157664   0.31805221] -400.0301095560293
